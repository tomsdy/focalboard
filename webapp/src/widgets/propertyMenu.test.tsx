// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'
import {fireEvent, render} from '@testing-library/react'
import '@testing-library/jest-dom'
import {IntlProvider} from 'react-intl'

import PropertyMenu from './propertyMenu'

const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>

describe('widgets/PropertyMenu', () => {
    beforeEach(() => {
        // Quick fix to disregard console error when unmounting a component
        console.error = jest.fn()
        document.execCommand = jest.fn()
    })

    test('should display the type of property', () => {
        const callback = jest.fn()
        const component = wrapIntl(
            <PropertyMenu
                propertyId={'id'}
                propertyName={'email of a person'}
                propertyType={'email'}
                onTypeChanged={callback}
                onNameChanged={callback}
                onDelete={callback}
            />,
        )
        const {getByText} = render(component)
        expect(getByText('Type: Email')).toBeVisible()
    })

    test('handles delete event', () => {
        const callback = jest.fn()
        const component = wrapIntl(
            <PropertyMenu
                propertyId={'id'}
                propertyName={'email of a person'}
                propertyType={'email'}
                onTypeChanged={callback}
                onNameChanged={callback}
                onDelete={callback}
            />,
        )
        const {getByText} = render(component)
        fireEvent.click(getByText(/delete/i))
        expect(callback).toHaveBeenCalledWith('id')
    })

    test('handles name change event', () => {
        const callback = jest.fn()
        const component = wrapIntl(
            <PropertyMenu
                propertyId={'id'}
                propertyName={'test-property'}
                propertyType={'text'}
                onTypeChanged={callback}
                onNameChanged={callback}
                onDelete={callback}
            />,
        )
        const {getByDisplayValue} = render(component)
        const input = getByDisplayValue(/test-property/i)
        fireEvent.change(input, {target: {value: 'changed name'}})
        fireEvent.blur(input)
        expect(callback).toHaveBeenCalledWith('changed name')
    })

    test('handles type change event', () => {
        const callback = jest.fn()
        const component = wrapIntl(
            <PropertyMenu
                propertyId={'id'}
                propertyName={'test-property'}
                propertyType={'text'}
                onTypeChanged={callback}
                onNameChanged={callback}
                onDelete={callback}
            />,
        )
        const {getByText} = render(component)
        const menuOpen = getByText(/Type: Text/i)
        fireEvent.click(menuOpen)
        fireEvent.click(getByText('Select'))
        expect(callback).toHaveBeenCalledWith('select')
    })

    test('should match snapshot', () => {
        const callback = jest.fn()
        const component = wrapIntl(
            <PropertyMenu
                propertyId={'id'}
                propertyName={'test-property'}
                propertyType={'text'}
                onTypeChanged={callback}
                onNameChanged={callback}
                onDelete={callback}
            />,
        )
        const {container, getByText} = render(component)
        const menuOpen = getByText(/Type: Text/i)
        fireEvent.click(menuOpen)
        expect(container).toMatchSnapshot()
    })
})
