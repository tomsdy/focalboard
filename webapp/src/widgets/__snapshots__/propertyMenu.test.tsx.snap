// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/PropertyMenu should match snapshot 1`] = `
<div>
  <div
    class="Menu noselect bottom"
  >
    <div
      class="menu-contents"
    >
      <div
        class="menu-options"
      >
        <input
          class="PropertyMenu menu-textbox"
          type="text"
          value="test-property"
        />
        <div
          class="MenuOption SubMenuOption menu-option"
          id="type"
        >
          <div
            class="noicon"
          />
          <div
            class="menu-name"
          >
            Type: Text
          </div>
          <svg
            class="SubmenuTriangleIcon Icon"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="50,35 75,50 50,65"
            />
          </svg>
          <div
            class="SubMenu Menu noselect bottom"
          >
            <div
              class="menu-contents"
            >
              <div
                class="menu-options"
              >
                <div
                  class="MenuOption LabelOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    <b>
                      Change property type
                    </b>
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="MenuOption MenuSeparator menu-separator"
                />
                <div
                  class="MenuOption TextOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    Text
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="MenuOption TextOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    Number
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="MenuOption TextOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    Email
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="MenuOption TextOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    URL
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="MenuOption TextOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    Select
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="MenuOption TextOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    Created Time
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
                <div
                  class="MenuOption TextOption menu-option"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    Updated Time
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
              </div>
              <div
                class="menu-spacer hideOnWidescreen"
              />
              <div
                class="menu-options hideOnWidescreen"
              >
                <div
                  class="MenuOption TextOption menu-option menu-cancel"
                >
                  <div
                    class="noicon"
                  />
                  <div
                    class="menu-name"
                  >
                    Cancel
                  </div>
                  <div
                    class="noicon"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MenuOption TextOption menu-option"
        >
          <div
            class="noicon"
          />
          <div
            class="menu-name"
          >
            Delete
          </div>
          <div
            class="noicon"
          />
        </div>
      </div>
      <div
        class="menu-spacer hideOnWidescreen"
      />
      <div
        class="menu-options hideOnWidescreen"
      >
        <div
          class="MenuOption TextOption menu-option menu-cancel"
        >
          <div
            class="noicon"
          />
          <div
            class="menu-name"
          >
            Cancel
          </div>
          <div
            class="noicon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
